File: 2470_number_of_subarrays_with_lcm_equal_to_k/solution.rb
# LeetCode 2470 - Number of Subarrays With LCM Equal to K
# https://leetcode.com/problems/number-of-subarrays-with-lcm-equal-to-k/

# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer}
def subarray_lcm(nums, k)
  gcd = lambda do |a, b|
    while b != 0
      a, b = b, a % b
    end
    a
  end

  ans = 0
  n = nums.length
  (0...n).each do |i|
    cur = 1
    (i...n).each do |j|
      cur = (cur / gcd.call(cur, nums[j])) * nums[j]
      break if cur > k

      ans += 1 if cur == k
    end
  end
  ans
end

File: 2470_number_of_subarrays_with_lcm_equal_to_k/solution.ts
// LeetCode 2470 - Number of Subarrays With LCM Equal to K
// https://leetcode.com/problems/number-of-subarrays-with-lcm-equal-to-k/

export function subarrayLCM(nums: number[], k: number): number {
    const gcd = (a, b) => {
        while (b !== 0) {
            const t = a % b;
            a = b;
            b = t;
        }
        return a;
    };
    let ans = 0;
    const n = nums.length;
    for (let i = 0; i < n; i++) {
        let cur = 1;
        for (let j = i; j < n; j++) {
            cur = Math.floor(cur / gcd(cur, nums[j])) * nums[j];
            if (cur > k) break;
            if (cur === k) ans++;
        }
    }
    return ans;
}
